File: app/src/main/java/ru/art2000/calculator/MainActivity.java
package ru.art2000.calculator;

import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.Toast;

import androidx.annotation.ColorInt;
import androidx.fragment.app.Fragment;
import androidx.viewpager2.widget.ViewPager2;

import com.sothree.slidinguppanel.SlidingUpPanelLayout;

import java.util.List;
import java.util.Objects;

import ru.art2000.calculator.calculator.CalculatorFragment;
import ru.art2000.calculator.currency_converter.CurrencyConverterFragment;
import ru.art2000.calculator.settings.SettingsFragment;
import ru.art2000.calculator.unit_converter.UnitConverterFragment;
import ru.art2000.extensions.DayNightActivity;
import ru.art2000.extensions.ReplaceableBottomNavigation;
import ru.art2000.extensions.ScrollControlledViewPager;
import ru.art2000.helpers.AndroidHelper;
import ru.art2000.helpers.CurrencyValuesHelper;
import ru.art2000.helpers.PrefsHelper;

public class MainActivity extends DayNightActivity {

    boolean useViewPager = true;
    boolean useViewPager2 = false;
    private boolean doubleBackToExitPressedOnce = false;
    private ReplaceableBottomNavigation navigation;
    private Context mContext;
    private CurrencyConverterFragment currency_converter;
    private CalculatorFragment calculator;
    private UnitConverterFragment unit_converter;
    private SettingsFragment settings;
    @ColorInt
    private int normalStatusBarColor;
    @ColorInt
    private int calculatorStatusBarColor;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        mContext = this;
        PrefsHelper.initialSetup(mContext);
        setTheme(PrefsHelper.getAppTheme());
        new Thread(() -> CurrencyValuesHelper.getDataFromDB(mContext)).start();
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        List<Fragment> list = getSupportFragmentManager().getFragments();
        for (Fragment fragment : list) {
            if (fragment instanceof CurrencyConverterFragment) {
                currency_converter = (CurrencyConverterFragment) fragment;
            } else if (fragment instanceof CalculatorFragment) {
                calculator = (CalculatorFragment) fragment;
            } else if (fragment instanceof UnitConverterFragment) {
                unit_converter = (UnitConverterFragment) fragment;
            } else if (fragment instanceof SettingsFragment) {
                settings = (SettingsFragment) fragment;
            }
        }
        if (currency_converter == null) {
            currency_converter = new CurrencyConverterFragment();
        }
        if (calculator == null) {
            calculator = new CalculatorFragment();
        }
        if (unit_converter == null) {
            unit_converter = new UnitConverterFragment();
        }
        if (settings == null) {
            settings = new SettingsFragment();
        }

        normalStatusBarColor = AndroidHelper.getColorAttribute(this, R.attr.colorPrimaryDark);
        calculatorStatusBarColor =
                AndroidHelper.getColorAttribute(this, R.attr.calc_input_bg);

        navigation = findViewById(R.id.navigation);

        navigation.setOnNavigationItemReselectedListener(item -> {
            switch (item.getItemId()) {
                case R.id.navigation_currency:
                    currency_converter.scrollToTop();
                    break;
                case R.id.navigation_calc:
                    SlidingUpPanelLayout.PanelState state = SlidingUpPanelLayout.PanelState.EXPANDED;
                    if (calculator.panel.getPanelState() == state)
                        state = SlidingUpPanelLayout.PanelState.COLLAPSED;
                    calculator.panel.setPanelState(state);
                    break;
            }
        });

        navigation.setOnNavigationItemSelectedListener(item -> {
            PrefsHelper.setDefaultTab(this, item.getOrder());

            switch (item.getItemId()) {
                default:
                case R.id.navigation_calc:
                    getIntent().setAction("ru.art2000.calculator.action.CALCULATOR");
                    break;
                case R.id.navigation_unit:
                    getIntent().setAction("ru.art2000.calculator.action.CONVERTER");
                    break;
                case R.id.navigation_currency:
                    getIntent().setAction("ru.art2000.calculator.action.CURRENCIES");
                    break;
                case R.id.navigation_settings:
                    getIntent().setAction("ru.art2000.calculator.action.SETTINGS");
                    break;
            }
            return true;
        });

        FrameLayout frameLayout = findViewById(R.id.fragment_container);
        ScrollControlledViewPager pager = findViewById(R.id.pager);
        ViewPager2 pager2 = findViewById(R.id.pager2);

        if (useViewPager2) {
            frameLayout.setVisibility(View.GONE);
            pager.setVisibility(View.GONE);

            navigation.setupWithViewPager2(
                    this,
                    pager2,
                    unit_converter, currency_converter, calculator, settings);

        } else if (useViewPager) {
            frameLayout.setVisibility(View.GONE);
            pager2.setVisibility(View.GONE);

            navigation.setupWithViewPager(
                    this,
                    pager,
                    unit_converter, currency_converter, calculator, settings);

        } else {
            pager.setVisibility(View.GONE);
            pager2.setVisibility(View.GONE);

            navigation.setupWithFragments(this,
                    R.id.fragment_container,
                    currency_converter, calculator, unit_converter, settings);

        }

        int tabId;
        if (Objects.requireNonNull(getIntent().getAction()).equals("android.intent.action.MAIN")) {
            tabId = PrefsHelper.getDefaultNavItem();
        } else {
            switch (getIntent().getAction()) {
                default:
                case "ru.art2000.calculator.action.CALCULATOR":
                    tabId = R.id.navigation_calc;
                    break;
                case "ru.art2000.calculator.action.CONVERTER":
                    tabId = R.id.navigation_unit;
                    break;
                case "ru.art2000.calculator.action.CURRENCIES":
                    tabId = R.id.navigation_currency;
                    break;
                case "ru.art2000.calculator.action.SETTINGS":
                    tabId = R.id.navigation_settings;
                    break;
            }
        }
        navigation.setSelectedItemId(tabId);
    }

    @Override
    protected void onPause() {
        if (currency_converter.adapter != null)
            currency_converter.adapter.removeEditText2();
        super.onPause();
    }

    @Override
    public void onBackPressed() {
        if (navigation.getSelectedItemId() == R.id.navigation_calc
                && (calculator.panel.getPanelState() == SlidingUpPanelLayout.PanelState.EXPANDED ||
                calculator.panel.getPanelState() == SlidingUpPanelLayout.PanelState.ANCHORED)) {

            calculator.panel.setPanelState(SlidingUpPanelLayout.PanelState.COLLAPSED);
        } else {
            if (doubleBackToExitPressedOnce) {
                finish();
                return;
            }
            doubleBackToExitPressedOnce = true;
            Toast.makeText(mContext, R.string.twice_tap_exit, Toast.LENGTH_SHORT).show();
            new Handler().postDelayed(() ->
                    doubleBackToExitPressedOnce = false, 2000);
        }
    }

    public void updateUnitView() {
        unit_converter.regenerateAdapter();
    }

    public void changeStatusBarColor(boolean isCalculatorPage) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (isCalculatorPage) {
                getWindow().setStatusBarColor(calculatorStatusBarColor);
            } else {
                getWindow().setStatusBarColor(normalStatusBarColor);
            }
        }
    }
}

File: app/src/main/java/ru/art2000/calculator/main/MainActivity.kt
package ru.art2000.calculator.main

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.View
import by.kirich1409.viewbindingdelegate.CreateMethod
import by.kirich1409.viewbindingdelegate.viewBinding
import com.google.android.material.navigation.NavigationBarView
import dagger.hilt.android.AndroidEntryPoint
import ru.art2000.calculator.common.model.MainPage
import ru.art2000.calculator.common.preferences.MainTabData
import ru.art2000.calculator.common.view.AppActivity
import ru.art2000.calculator.databinding.ActivityMainBinding
import ru.art2000.calculator.settings.preferences.PreferenceValues
import ru.art2000.extensions.views.isLandscape
import ru.art2000.extensions.views.setMyOnItemSelectedListener
import ru.art2000.extensions.views.setSelectedItemId
import ru.art2000.extensions.views.setupWithViewPager2
import javax.inject.Inject

@AndroidEntryPoint
class MainActivity : AppActivity() {
    private val binding by viewBinding<ActivityMainBinding>(CreateMethod.INFLATE)

    @Inject
    lateinit var tabDatas: Map<MainPage, @JvmSuppressWildcards MainTabData<*>>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            startActivity(Intent(this, UpdateActivity::class.java))
            finish()
            return
        }

        setContentView(binding.root)

        val navigation = binding.navigation as NavigationBarView

        val (pages, orderedData) = tabDatas
            .toSortedMap(Comparator.comparing { it.ordinal })
            .let { it.keys.toList() to it.values }

        navigation.setMyOnItemSelectedListener { item ->

            val page = pages[orderedData.indexOfFirst { it.idRes == item.itemId }]!!

            generalPrefsHelper.defaultTabToOpen = when (page) {
                MainPage.UNIT -> PreferenceValues.VALUE_TAB_DEFAULT_UNIT
                MainPage.CURRENCY -> PreferenceValues.VALUE_TAB_DEFAULT_CURRENCY
                MainPage.SETTINGS -> PreferenceValues.VALUE_TAB_DEFAULT_SETTINGS
                MainPage.CALCULATOR -> PreferenceValues.VALUE_TAB_DEFAULT_CALC
            }
            intent.action = when (page) {
                MainPage.UNIT -> ACTION_CONVERTER
                MainPage.CURRENCY -> ACTION_CURRENCIES
                MainPage.SETTINGS -> ACTION_SETTINGS
                MainPage.CALCULATOR -> ACTION_CALCULATOR
            }

            true
        }

        navigation.setupWithViewPager2(
            this,
            binding.pager2,
            orderedData.map { it.tabCreator },
        )

        val page = when (intent.action) {
            Intent.ACTION_MAIN -> when (generalPrefsHelper.defaultTabToOpen) {
                PreferenceValues.VALUE_TAB_DEFAULT_UNIT -> MainPage.UNIT
                PreferenceValues.VALUE_TAB_DEFAULT_CURRENCY -> MainPage.CURRENCY
                PreferenceValues.VALUE_TAB_DEFAULT_SETTINGS -> MainPage.SETTINGS
                else -> MainPage.CALCULATOR
            }
            ACTION_CALCULATOR -> MainPage.CALCULATOR
            ACTION_CONVERTER -> MainPage.UNIT
            ACTION_CURRENCIES -> MainPage.CURRENCY
            ACTION_SETTINGS -> MainPage.SETTINGS
            else -> MainPage.CALCULATOR
        }
        navigation.setSelectedItemId(tabDatas[page]!!.idRes, false)
    }

    override val bottomViews: List<View>
        get() = if (isLandscape) emptyList() else listOf(binding.navigation)

    override val leftViews: List<View>
        get() = listOf(binding.root)

    override val rightViews: List<View>
        get() = if (isLandscape) listOf(binding.navigation) else listOf(binding.root)

    private companion object {

        const val ACTION_CALCULATOR = "ru.art2000.calculator.action.CALCULATOR"
        const val ACTION_CONVERTER = "ru.art2000.calculator.action.CONVERTER"
        const val ACTION_CURRENCIES = "ru.art2000.calculator.action.CURRENCIES"
        const val ACTION_SETTINGS = "ru.art2000.calculator.action.SETTINGS"

    }
}